plugins/techdocs-backend/src/techdocs/stages/generate/helpers.ts

```diff
@@ -142,3 +147,131 @@ export const runCommand = async ({
+ * @param {string} repoUrl URL supposed to be used as repo_url in mkdocs.yml
+ * @param {RemoteProtocol} locationType Type of source code host - github, gitlab, dir, url, etc.
+ * @returns {boolean}
+ */
+export const isValidRepoUrlForMkdocs = (
+  repoUrl: string,
+  locationType: RemoteProtocol,
+): boolean => {
+  // Trim trailing slash
+  const cleanRepoUrl = repoUrl.replace(/\/$/, '');
+
+  if (locationType === 'github' || locationType === 'gitlab') {
+    // A valid repoUrl to the root of the repository will be split into 5 strings if split using the / delimiter.
+    // We do not want URLs which have more than that number of forward slashes since they will signify a non-root location
+    // Note: This is not the best possible implementation but will work most of the times.. Feel free to improve or
+    // highlight edge cases.
+    return cleanRepoUrl.split('/').length === 5;
+  }
+
+  return false;
+};
+
+/**
+ * Return a valid URL of the repository used in backstage.io/techdocs-ref annotation.
+ * Return undefined if the `target` is not valid in context of repo_url in mkdocs.yml
+ * Alter URL so that it is a valid repo_url config in mkdocs.yml
+ *
+ * @param {ParsedLocationAnnotation} parsedLocationAnnotation Object with location url and type
+ * @returns {string | undefined}
+ */
+export const getRepoUrlFromLocationAnnotation = (
+  parsedLocationAnnotation: ParsedLocationAnnotation,
+): string | undefined => {
+  const { type: locationType, target } = parsedLocationAnnotation;
+
+  // Add more options from the RemoteProtocol type of parsedLocationAnnotation.type here
+  // when TechDocs supports more hosts and if mkdocs can generated an Edit URL for them.
+  const supportedHosts = ['github', 'gitlab'];
+
+  if (supportedHosts.includes(locationType)) {
+    // Trim .git or .git/ from the end of repository url
+    return target.replace(/.git\/*$/, '');
+  }
+
+  return undefined;
+};
+
+/**
+ * Update the mkdocs.yml file before TechDocs generator uses it to build docs site.
+ *
+ * List of tasks:
+ * - Add repo_url if it does not exists
+ * If mkdocs.yml has a repo_url, the generated docs site gets an Edit button on the pages by default.
+ * If repo_url is missing in mkdocs.yml, we will use techdocs annotation of the entity to possibly get
+ * the repository URL.
+ *
+ * This function will not throw an error since this is not critical to the whole TechDocs pipeline.
+ * Instead it will log warnings if there are any errors in reading, parsing or writing YAML.
+ *
+ * @param {string} mkdocsYmlPath Absolute path to mkdocs.yml or equivalent of a docs site
+ * @param {Logger} logger
+ * @param {ParsedLocationAnnotation} parsedLocationAnnotation Object with location url and type
+ */
+export const patchMkdocsYmlPreBuild = async (
+  mkdocsYmlPath: string,
+  logger: Logger,
+  parsedLocationAnnotation: ParsedLocationAnnotation,
+) => {
+  let mkdocsYmlFileString;
+  try {
+    mkdocsYmlFileString = fs.readFileSync(mkdocsYmlPath, 'utf8');
```

Comment:
We're in an async function already, don't use the Sync methods; use fs-extra. Same below